### PR TITLE
Remove LinkedDeque and replace with LinkedHashMap

### PR DESCRIPTION
**Pull request by andrross**
After the recent changes the usage of the LinkedDeque fits quite well with the insertion order semantics of LinkedHashMap, which also allows for constant time additions and removals.

### Microbenchmark Results

The benchmark results are quite similar. The LinkedHashMap version seems to be slightly faster, but they are close enough there could be other variables in the test runs causing the difference. The benchmarks show there isn't any significant regression.

OS: Ubuntu 20.04.5 LTS
Host: AWS EC2 c6i.8xlarge
Benchmark command: `./gradlew -p benchmarks run --args 'FileCacheBenchmark'`

This PR:
```
Benchmark                   (concurrencyLevel)  (maximumNumberOfEntries)   Mode  Cnt      Score   Error   Units
FileCacheBenchmark.get                       1                     65536  thrpt        3395.869          ops/ms
FileCacheBenchmark.get                       1                   1048576  thrpt        1542.995          ops/ms
FileCacheBenchmark.get                       8                     65536  thrpt        6674.145          ops/ms
FileCacheBenchmark.get                       8                   1048576  thrpt        3810.060          ops/ms
FileCacheBenchmark.put                       1                     65536  thrpt        3608.218          ops/ms
FileCacheBenchmark.put                       1                   1048576  thrpt        1667.715          ops/ms
FileCacheBenchmark.put                       8                     65536  thrpt        7229.409          ops/ms
FileCacheBenchmark.put                       8                   1048576  thrpt        4456.812          ops/ms
FileCacheBenchmark.remove                    1                     65536  thrpt       12417.484          ops/ms
FileCacheBenchmark.remove                    1                   1048576  thrpt        8151.180          ops/ms
FileCacheBenchmark.remove                    8                     65536  thrpt       18792.732          ops/ms
FileCacheBenchmark.remove                    8                   1048576  thrpt       15913.771          ops/ms
FileCacheBenchmark.replace                   1                     65536  thrpt        3326.174          ops/ms
FileCacheBenchmark.replace                   1                   1048576  thrpt        1485.508          ops/ms
FileCacheBenchmark.replace                   8                     65536  thrpt        6783.510          ops/ms
FileCacheBenchmark.replace                   8                   1048576  thrpt        3947.464          ops/ms
```

Previous implementation:
```
Benchmark                   (concurrencyLevel)  (maximumNumberOfEntries)   Mode  Cnt      Score   Error   Units
FileCacheBenchmark.get                       1                     65536  thrpt        3344.929          ops/ms
FileCacheBenchmark.get                       1                   1048576  thrpt        1513.615          ops/ms
FileCacheBenchmark.get                       8                     65536  thrpt        6536.668          ops/ms
FileCacheBenchmark.get                       8                   1048576  thrpt        3869.711          ops/ms
FileCacheBenchmark.put                       1                     65536  thrpt        3704.204          ops/ms
FileCacheBenchmark.put                       1                   1048576  thrpt        1684.997          ops/ms
FileCacheBenchmark.put                       8                     65536  thrpt        7023.869          ops/ms
FileCacheBenchmark.put                       8                   1048576  thrpt        4276.159          ops/ms
FileCacheBenchmark.remove                    1                     65536  thrpt       12547.969          ops/ms
FileCacheBenchmark.remove                    1                   1048576  thrpt        8131.708          ops/ms
FileCacheBenchmark.remove                    8                     65536  thrpt       16763.030          ops/ms
FileCacheBenchmark.remove                    8                   1048576  thrpt       15659.768          ops/ms
FileCacheBenchmark.replace                   1                     65536  thrpt        3394.281          ops/ms
FileCacheBenchmark.replace                   1                   1048576  thrpt        1495.073          ops/ms
FileCacheBenchmark.replace                   8                     65536  thrpt        7018.876          ops/ms
FileCacheBenchmark.replace                   8                   1048576  thrpt        3796.394          ops/ms
```

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by reta** [comment]
@andrross I like it, the implementation is much simpler now, interestingly enough, the `LinkedHashMap` supports eviction with `removeEldestEntry` method (that could be implemented) but since we use 2 maps (`data` and `lru` that should be in sync), we may not benefit from it.

File: server/src/main/java/org/opensearch/index/store/remote/utils/cache/LRUCache.java [hunk 1]
```diff
@@ -45,15 +47,15 @@
...
+    private final LinkedHashMap<K, Node<K, V>> lru;
 
     private final RemovalListener<K, V> listener;
 
     private final Weigher<V> weigher;
 
     private final StatsCounter<K> statsCounter;
 
-    private volatile ReentrantLock lock;
+    private final ReentrantLock lock;
```
Comment (reta) [inline server/src/main/java/org/opensearch/index/store/remote/utils/cache/LRUCache.java:58]: :+1:  we could replace

```
        final ReentrantLock lock = this.lock;
        lock.lock();
```

with just

```
        lock.lock();
```

now

**Review by reta — APPROVED** [review]

File: server/src/main/java/org/opensearch/index/store/remote/utils/cache/LRUCache.java (hunk 1 above)
Comment (andrross) [reply server/src/main/java/org/opensearch/index/store/remote/utils/cache/LRUCache.java:58]: Oh good catch! I'm sure there was some previous incarnation of this code where a volatile lock was needed, but it definitely was never necessary here.

**Comment by andrross** [comment]
@reta regarding `removeEldestEntry`/cache improvements in general, I added some more details here with where the cache has evolved since that issue was opened, with my opinions on what should be improved: https://github.com/opensearch-project/OpenSearch/issues/6225#issuecomment-1496337366

**Comment by kotwanikunal** [comment]
@andrross Do we need to backport this to `2.x`? Backport for #7498 got me here since there was a conflict

**Comment by dblock** [comment]
Looks like https://github.com/opensearch-project/OpenSearch/pull/7595 includes this? @kotwanikunal

**Comment by kotwanikunal** [comment]
> Looks like #7595 includes this? @kotwanikunal

Yes @dblock. Backport was done with #7595

**Pull request merged**